src/back/msl/writer.rs

```diff
@@ -1061,25 +1063,37 @@ impl<W: Write> Writer<W> {
             // has to be a named expression
             crate::Expression::Call(_) => unreachable!(),
             crate::Expression::ArrayLength(expr) => match *context.resolve_type(expr) {
-                crate::TypeInner::Array {
-                    size: crate::ArraySize::Constant(const_handle),
-                    ..
-                } => {
-                    let coco = ConstantContext {
-                        handle: const_handle,
-                        arena: &context.module.constants,
-                        names: &self.names,
-                        first_time: false,
-                    };
-                    write!(self.out, "{}", coco)?;
-                }
-                crate::TypeInner::Array { .. } => {
-                    return Err(Error::FeatureNotImplemented(
-                        "dynamic array size".to_string(),
-                    ))
-                }
+                crate::TypeInner::Pointer { base, .. } => match context.module.types[base].inner {
+                    crate::TypeInner::Array { stride, .. } => {
+                        let handle = context.info[expr].assignable_global.unwrap();
+
+                        let global = &context.module.global_variables[handle];
+                        if let crate::TypeInner::Struct { ref members, .. } = context.module.types[global.ty].inner {
+                            if let Some(&crate::StructMember { offset, .. }) = members.last() {
+                                let buffer_idx = self.runtime_sized_buffers[&handle];
+                                write!(self.out, "((x_buffer_sizes.buffer_size{} - {}) / {})", buffer_idx, offset, stride)?;
```

Comment:
This may be a little bit more tricky. The stride can be larger than the size of the type. So we may end up in a situation where there is space for N+1 elements, but dividing by the stride will give you N.
You can get the span occupied by this type via `context.module.types[member.y].inner.span(&module.constants)`.

One of the ways to write the proper equation would be :
```
write!(self.out, "(1 + (x_buffer_sizes.buffer_size{} - {} - {}) / {})", buffer_idx, offset, type_span, stride)?;
```
(where the `type_span` is the span for `member.ty`).
Note: this array is guaranteed to have at least one element.